frr-tools/status/controller/bgpsessionstate_controller.go

```diff
@@ -0,0 +1,268 @@
+				vrfLabel:  vrf,
+			}
+			bfdStatus := ""
+			if _, ok := bfdForPeer[vrf]; ok {
+				bfdStatus = bfdForPeer[vrf][neigh.ID]
+			}
+			if bfdStatus == "" {
+				bfdStatus = noBFDConfigured
+			}
+			desiredState.Status = frrk8sv1beta1.BGPSessionStateStatus{
+				Node:      r.NodeName,
+				Peer:      labelFormatForNeighbor(neigh.ID),
+				VRF:       vrf,
+				BGPStatus: neigh.BGPState,
+				BFDStatus: bfdStatus,
+			}
+
+			if existingState != nil && reflect.DeepEqual(desiredState.Labels, existingState.Labels) && reflect.DeepEqual(desiredState.Status, existingState.Status) {
+				continue
+			}
+			toApply[vrf][neigh.ID] = desiredState
+		}
+	}
+
+	errs := []error{}
+	for _, states := range existing { // delete the existing statuses that belong to non-existing neighbors
+		for _, s := range states {
+			s := s
+			err := r.Client.Delete(ctx, s)
+			if err != nil && !apierrors.IsNotFound(err) {
+				errs = append(errs, err)
+			}
+		}
+	}
+
+	for _, states := range toApply {
+		for _, s := range states {
+			s := s
+			desiredStatus := s.Status
+			_, err := controllerutil.CreateOrPatch(ctx, r.Client, s, func() error {
+				err = controllerutil.SetOwnerReference(r.DaemonPod, s, r.Scheme())
+				if err != nil {
+					return err
+				}
+				s.Status = desiredStatus
+				return nil
+			})
+			if err != nil {
+				errs = append(errs, err)
+			}
+		}
+	}
+
+	if utilerrors.NewAggregate(errs) != nil {
+		return ctrl.Result{}, utilerrors.NewAggregate(errs)
+	}
+
+	// We use the ResyncPeriod for requeuing the node's FRRNodeState, relying on it being
+	// the only non-namespaced resource with the node's name that triggers the reconciliation.
+	if req.Name == r.NodeName && req.Namespace == "" {
+		return ctrl.Result{RequeueAfter: r.ResyncPeriod}, nil
+	}
+
+	return ctrl.Result{}, nil
+}
+
+func (r *BGPSessionStateReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	p := predicate.NewPredicateFuncs(func(o client.Object) bool {
+		return r.filterBGPSessionStateEvent(o) && r.filterFRRNodeStateEvent(o)
+	})
+
+	return ctrl.NewControllerManagedBy(mgr).
+		For(&frrk8sv1beta1.BGPSessionState{}).
+		Watches(&frrk8sv1beta1.FRRNodeState{}, &handler.EnqueueRequestForObject{}).
+		WithEventFilter(p).
+		Complete(r)
+}
+
+func (r *BGPSessionStateReconciler) filterBGPSessionStateEvent(o client.Object) bool {
+	s, ok := o.(*frrk8sv1beta1.BGPSessionState)
+	if !ok {
+		return true
+	}
+
+	if s.Labels == nil {
+		return false
+	}
+
+	if nodeFor(*s) != r.NodeName {
+		return false
+	}
+
+	return true
+}
+
+func (r *BGPSessionStateReconciler) filterFRRNodeStateEvent(o client.Object) bool {
+	s, ok := o.(*frrk8sv1beta1.FRRNodeState)
+	if !ok {
+		return true
+	}
+
+	if s.Name != r.NodeName {
+		return false
+	}
+
+	return true
+}
+
+func nodeFor(s frrk8sv1beta1.BGPSessionState) string {
+	return s.Labels[nodeLabel]
+}
+
+func peerFor(s frrk8sv1beta1.BGPSessionState) string {
+	return s.Labels[peerLabel]
+}
+
+func vrfFor(s frrk8sv1beta1.BGPSessionState) string {
```

Comment:
Why we cannot use directly  `s.Status.VRF` ?

Reply:
same reason as https://github.com/metallb/metallb/pull/2650#discussion_r1933869077